### ISSUE TITLE
Pyright hangs on minimal data-science stack

### ISSUE THREAD

**Issue by randomgeek78** [microsoft/pyright#3206]
**Describe the bug**
Pyright hangs on my minimal data-science stack code base. It seems certain statements trigger the hang. Once pyright hangs, it becomes unresponsive and uses 100% CPU (forever).

**To Reproduce**
```
$ poetry new data-science
$ cd data-science
$ cat pyproject.toml
[tool.poetry]
name = "data-science"
version = "0.1.0"
description = ""
authors = ["Your Name <you@example.com>"]

[tool.poetry.dependencies]
python = "^3.8"
pandas = "^1.4.1"
scikit-learn = "^1.0.2"
numpy = "^1.22.3"

[tool.poetry.dev-dependencies]
pytest = "^5.2"
ipython = "^8.1.1"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

$ cat pyrightconfig.json
{
  "include": [
    "data_science"
  ],
  "executionEnvironments": [
    {
      "root": "data_science"
    }
  ],
  "venv": ".venv",
  "venvPath": "."
}

$ poetry config --list
cache-dir = "/home/popos/.cache/pypoetry"
experimental.new-installer = true
installer.parallel = true
repositories.testpypi.url = "https://test.pypi.org/legacy/"
virtualenvs.create = true
virtualenvs.in-project = true
virtualenvs.path = "{cache-dir}/virtualenvs"  # /home/popos/.cache/pypoetry/virtualenvs

$ poetry install # installs in ./.venv

$ cat example-1.py
from sklearn.decomposition import PCA
import numpy as np
from sklearn.cluster import KMeans

training_data = np.random.randn(100, 3)
pca = PCA(n_components=3).fit(training_data)
# kmeans = KMeans(init=pca.components_, n_clusters=3, n_init=1) #  # triggers hang

$ cat example-2.py
from sklearn.decomposition import TruncatedSVD
from sklearn.datasets import load_iris

iris = load_iris()
X, y = iris.data, iris.target

svd = TruncatedSVD(n_components=2, random_state=0)
# Xt = svd.fit_transform(X, y) # triggers hang
```

**Expected behavior**
In the examples 1 and 2 above, the last line triggers the hang. As mentioned in the beginning, this causes pyright to start consuming 100% CPU until the process is killed.

I know there are some other threads that also talk about high CPU usage and I was not sure this one is related. I wanted to provide a complete setup. I am still not sure if I am setting things up properly. If my setup can be changed so I can be unblocked, I would welcome any such suggestions.

**Comment by erictraut** [comment]
This is likely similar to (or maybe even the same issue as) https://github.com/microsoft/pyright/issues/3156. The sklearn library contains some functions that are very complex from a code flow standpoint, and none of the functions or variables are annotated, so all types need to be inferred.

I'll take a look at this when I have a chance. (Perf bugs like this tend to take a lot of time to diagnose.) In the meantime, the best workaround is to use a partial type stub for sklearn. Pylance recently added such stubs, and I've asked the pylance team to check them into [this public repo](https://github.com/microsoft/python-type-stubs) so you can access them.

If sklearn is a library that is important to you, I would suggest that you contact the maintainers and encourage them to add type declarations to the library. If they're unwilling to do this, you might consider creating typeshed type stubs for the library.

**Comment by erictraut** [comment]
Another workaround is to disable the `useLibraryCodeForTypes` setting, which is off by default in pyright. I presume you have enabled it (although I don't see that above). When this setting is disabled, pyright won't attempt to infer any types from an untyped library.

**Comment by randomgeek78** [comment]
Thanks @erictraut. Although I didn't know about this option earlier, I can confirm that I had useLibraryCodeForTypes set to true. When I set this option to false, I no longer have hangs with the above examples. I also noticed that with this option set to false, I am no longer able to jump to definition directly. 

```python
from sklearn.decomposition import PCA
pca = PCA(n_components=3).fit(training_data)
```
In the above, I am no longer to able to jump to PCA definition directly. But I am able to jump to definition of decomposition and then from there, I am able to jump to definition of PCA through 

```python
# sklearn/decomposition/__init__.py
...
from ._pca import PCA
...
```

I would like to confirm if useLibraryCodeForTypes is required to jump to definition directly. From the description, it was not clear to me what exactly is affected. 

Besides the pyrightconfig.json that I quoted, I also had autoSearchPaths = true, useLibraryCodeForTypes = false and diagnosticMode = 'workspace' in another place.

**Comment by erictraut** [comment]
When `useLibraryCodeForTypes` is off, pyright doesn't attempt to analyze any of the code in the library. It assumes all imported symbols have the type `Unknown` (which is a special form of `Any`). That means all type knowledge — including declaration and definition sites — are unknown for that symbol.

As I said, the real solution here is for the maintainers of sklearn to provide inlined type annotations for their public interface surface. We've provided [this guidance](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) for library authors. If they're unwilling to do add type information, then the community can fill in the missing information through external type stubs.

**Comment by randomgeek78** [comment]
Thank you.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.234, which I just published. It will also be included in the next pylance release.
